gen3_tracker/git/adder.py

```diff
@@ -0,0 +1,302 @@
+import glob
+import pathlib
+import sys
+from datetime import datetime
+from urllib.parse import urlparse, ParseResult
+
+import click
+import dateutil
+import pytz
+import yaml
+from dateutil.tz import tzutc
+
+from gen3_tracker.common import ACCEPTABLE_HASHES
+
+
+def is_valid_url(target):
+    try:
+        result = urlparse(target)
+        return all([result.scheme, result.netloc])
+    except ValueError:
+        return False
+
+
+# handle http, https
+def get_http_cloud_storage_path(url):
+    parsed_url = urlparse(url)
+    if parsed_url.scheme not in ['http', 'https']:
+        raise ValueError(f'Invalid scheme {parsed_url.scheme} for cloud storage URL')
+
+    # The path component of the parsed URL starts with a '/', so we remove it with [1:]
+    # For MinIO, the path includes the bucket name, so we need to split on '/' and take everything after the first element
+    return '/'.join(parsed_url.path.split('/')[2:])
+
+
+def get_http_bucket_from_url(url):
+    parsed_url = urlparse(url)
+    if parsed_url.scheme not in ['http', 'https']:
+        raise ValueError(f'Invalid scheme {parsed_url.scheme} for cloud storage URL')
+
+    # For MinIO, the bucket name is the first element in the path
+    return parsed_url.path.split('/')[1]
+
+
+# handle s3, gs, azure
+def get_cloud_storage_path(url) -> ParseResult:
+    """breaks the URL into components, which can then be used to derive the file path."""
+    parsed_url: ParseResult = urlparse(url)
+    if parsed_url.scheme not in ['s3', 'gs', 'azure']:
+        raise ValueError(f'Invalid scheme {parsed_url.scheme} for cloud storage URL')
+
+    # The path component of the parsed URL starts with a '/', so we remove it with [1:]
+    return parsed_url
+
+
+def add_url(ctx, target) -> tuple[list[pathlib.Path], list[str]]:
+    from gen3_tracker.common import ACCEPTABLE_HASHES, INFO_COLOR
+
+    """Add a url to the repository. If the file is already in the repository, it will be updated."""
+    assert is_valid_url(target), f'{target} is not a valid url.'
+    url = target
+    from gen3_tracker.git import git_files
+
+    files_already_in_repo = git_files()
+    updates = []  # only updates
+    all_changed_files = []  # new and updates
+
+    target = url_path(url)
+
+    if f"MANIFEST/{target}.dvc" in files_already_in_repo:
+        click.secho(f'{target} is already in the repository. updating.', fg=INFO_COLOR, file=sys.stderr)
```

Comment:
Info level click.secho commands going to sys.stderr and not sys.stdout ?

Reply:
Yes,  in that way  informative text messages can be displayed without interfering with yaml or json going to stdout.  Makes it easier to use in a tool chain.